### ISSUE TITLE
Add support to detect and install VSIX packages into VS15

### ISSUE THREAD

**Issue by heaths** [wixtoolset/issues#5433]
* Describe the scenario and benefits that the feature supports.

The detection properties based on registry or component searches for VS15 and support to install VSIX packages via the `VsixPackage` will no longer work for VS15. The setup engine has been completely redesigned to use a new VSIX format extension that provides a number of benefits, mostly notably more flexibility and potentially multiple installs. As such, support for VS15 needs to be added to continue supporting detection properties and `VsixPackage` support.

I have already written and published official native and managed (via RCW embeddable types) COM servers via NuGet, described [here](https://blogs.msdn.microsoft.com/heaths/2016/09/15/changes-to-visual-studio-15-setup).

* Describe how you're accomplishing the feature today (if possible).

Past releases used singleton installers via MSI and could write registry values for detection. Additionally, much of the product was installed invariably and extension developers could typically rely on any feature on which they depend to be installed. That is no longer the case. Right now, even if you did know exactly where to find _VsixInstaller.exe_, there is no way to know for sure it will work if it has dependencies on things not installed.

* Describe what you'd like the new feature to do.

Ideally, many extension developers or those just looking for where VS is installed do not have to change their authoring style. With custom actions that call into the COM servers, location properties can be set to the newest VS15 installed meaning their extension would install into that location.

However, because multiple instances can be installed and with different feature sets, the existing schema could be extended to support:

1. To set the sole `VS15_ROOT_FOLDER` an optional set of workloads or components can be specified to make sure the instance of VS used supports those.
2. To install a `VsixPackage` an optional set of workloads or components can be specified to make sure the VSIX package would be supported.
3. Provide install time support to install an extension to multiple instances supporting an optional set of workloads or components.

**Comment by heaths** [comment]
I'm working on the WIP for this and met to discuss certain scenarios with the VS Extensibility team today, so this issue will not stand alone for much longer.

**Comment by justcla** [comment]
Thanks for looking into this. Having support for Dev15 (and it's multiple-instance model) would make a huge difference to many extension developers.

**Comment by heaths** [comment]
Seems only fitting. I helped write some of WiX, have updated the VSExtension for the past releases, and designed much of the new engine and wrote the query APIs. :)

Because of the multi-instance model, the properties will be less useful but we may still keep most of them (I just met with the Extensibility team yesterday to discuss these). While I'll call it out in the WIP, I'll likely point them at the newest install of the highest version available. But the `VsixPackage` should work for multiple instances. It's just that the properties can only point at one directory (per property).

**Comment by niemyjski** [comment]
I'm not sure why it's not possible to just add support for the new VsixPackage features but for now at least update WIX to include support for all the traditional VS properties. We will still try to maintain creating shortcuts to package defs (if you can still do it in VS2017) as we do to support 2010-2017

**Comment by heaths** [comment]
There are no searches supported that would find instances of VS15. You can't find them via registry, components (in fact, we use almost no MSIs anymore), or a simple directory/file search (because you wouldn't know from where to start).

I will be ready with the associated WIP soon that will go into details about how I'm planning to address this using the query API I described in my blog post.

**Comment by niemyjski** [comment]
@heaths is there any chance we could meet up, I'd love to show you how we are using WIX to install a commercial (single) VSPackage into VS2010,2012,2013 & 2015. Would love to get your feedback on this and show you our scenario.

**Comment by mletterle** [comment]
I, too, am anxiously awaiting this support, has there been any progress? Willing to help if needed!

**Comment by heaths** [comment]
There is a thread on the wix-devs distribution list discussing options. Please see http://wixtoolset.org/documentation/ for how to join.

**Comment by NetTecture** [comment]
Please consider integrating WIX a lot more with the way .NET works. Split it.

* A visual studio extension that is published and maintained through the market place for the visual studio part (editor)
* A nuget based tool distribution for the tools.

The advantage would be in VS 2017:

* Automatic installation of WIX on multiple VS Installs (through extension synchronization between machines) for a developer.
* No need to maintain anything on a build server as the nuget restore would handle that.

Result: significanly easier handling.

**Comment by robmen** [comment]
@NetTecture there are plans for that in WiX v4. This issue is about enhancing existing the support in the WiX Toolset to support VS2017.

**Comment by NetTecture** [comment]
@robmen Lots of people do use it today in v3. Not delivering nuget binaries means they have to resoirt to third party tools which may or may not be updated properly. Even if v4 comes out they will not switch over immedately. If this thread is about enchancing the existing support, then consider doing it so it is in line with the guidelines of the tools. RIght now I am using v3 in some projects and to use only official sources I am forced to install visual studio on a build server, while the stack perfectly works without it. Or, as I said, use non-official package sources. Particularyly in the VS2017 frame, I would assume many pople will want to move towards NOT installing visual studio on a build server.

To my other topic: delivering the experience as visual studio extension contrary to an indendent installer makes it significantly more useful. I work on 4 installations of visual studio in parallel (different vm's for different customers), and WIX is a baseline component. VS 2017 allows me to keep those synchronized. The existing installer forces me so to say to regularly check manally for a new release (which is not that bad becasuse by now they rarely happen due to the stability).

**Comment by robmen** [comment]
@NetTecture a few things:

1. This issue tracks the work to update the WixVsExtension to support VS2017. Discussions about the "enhancing the existing support" take place on wix-devs mailing list.

2. There is no requirement to install Visual Studio on a build machine for the WiX Toolset. AFAIK, there never has been.

3. For all the rest of your completely valid points we are looking to address them in WiX v4. Our focus will completely shift to v4 when v3.11 is complete (aka: very soon).

Finally, happy to continue this conversation on the wix-devs mailing list but it does not belong on this GitHub issue. It turns out implementing VS2017 support in the WixVsExtension is complex enough without bringing in aspirations for other feature work. :smile:

**Comment by barnson** [comment]
@heaths: Will the detection code be available for WiX v3.11? Even without the invocation of VSIXInstaller, detection would be useful.

**Comment by heaths** [comment]
Can you describe a scenario? Without the ability to put anything into instances of VS, what good is detection? I can try to finish it up in time (how much?) for 3.11 if there’s sufficient reason.

**Comment by barnson** [comment]
UI: "We noticed you have VS2017; go <here> blah-blah-blah." See also https://github.com/wixtoolset/issues/issues/5510.

**Comment by heaths** [comment]
Okay. But do you need the properties as I described them in the WIP hooked up? Even them I’m thinking they don’t make much sense. I could easily write something that tells you just that VS2017 is installed (providing version-specific variables on the same code base).

What is the release timeline for 3.11? I’ll prioritize whatever we decide (i.e. original properties or just an existence check) accordingly.

**Comment by barnson** [comment]
Existence would be pri0. Version: pri1. A display name would be pri0.5.

RC happens tomorrow. RTW in a few weeks, depending on RC feedback.

**Comment by heaths** [comment]
Looking at the related work item regarding the VS marketplace, if Votive has no dependencies other than CoreEditor (I’d be surprised if it – or most extensions, for that matter – did) the concerns raised in the related WIP shouldn’t be a concern. Only optional, not-installed workload or components could cause the nested MSI problem.

Once the changes to VSIXInstaller.exe are in place, Votive could once again be installed with the MSI. You could also hook up https://github.com/Microsoft/vsixbootstrapper to the bundle if you wanted to now. But there’s really no way to do a DetectCondition so it’d have to run every time (but should basically be a no-op subsequently).

**Comment by barnson** [comment]
How can we detect that the fixed VSIXInstaller is on the machine? Version number?

**Comment by heaths** [comment]
Correct. VSIXInstaller.exe is already shipping in/with the setup engine but won’t have that functionality at RTW.

**Comment by barnson** [comment]
So we'll need the path to do a version check. Pri2 but should be straightforward since the CA already needs the path, no?

**Comment by heaths** [comment]
Yeah.

**Comment by pkeuter** [comment]
I am not really sure how related this is. Maybe it isn't, in that case I can make a separate issue for this. But after installing VS15 (+Votive2017 and 3.10 or 3.11), my Custom Actions project fails to load. It is looking for a Wix.CA.targets file, which it cannot find. This happens when you create a CA project from scratch, so it's easily reproducible.

**Comment by robmen** [comment]
@pkeuter Not related. See #5511.

**Comment by pkeuter** [comment]
@robmen, 7 minutes ago. There is a reason I missed that issue ;-) Thanks, I'll follow that.

**Comment by robmen** [comment]
Awesome, look forward to future improvements that will add the ability to *install* as well. 😄

**Comment by niemyjski** [comment]
Any updates on this?

**Comment by jlgrau** [comment]
Hello do you have any sample project to explain how to build vsixbootstrapper.exe tool?

**Comment by heaths** [comment]
@jlgrau check out the project site: https://github.com/Microsoft/vsixbootstrapper.
